blob/memblob/memblob.go

```diff
@@ -0,0 +1,336 @@
+	})
+}
+
+// Options sets options for constructing a *blob.Bucket backed by memory.
+type Options struct{}
+
+type blobEntry struct {
+	Content    []byte
+	Attributes driver.Attributes
+}
+
+type bucket struct {
+	mu    sync.Mutex
+	blobs map[string]*blobEntry
+}
+
+// openBucket creates a driver.Bucket backed by memory.
+func openBucket(_ *Options) driver.Bucket {
+	return &bucket{
+		blobs: map[string]*blobEntry{},
+	}
+}
+
+// OpenBucket creates a *blob.Bucket backed by memory.
+func OpenBucket(opts *Options) *blob.Bucket {
+	return blob.NewBucket(openBucket(opts))
+}
+
+// IsNotExist implements driver.IsNotExist.
+func (b *bucket) IsNotExist(err error) bool {
+	return err == errNotFound
+}
+
+// IsNotImplemented implements driver.IsNotImplemented.
+func (b *bucket) IsNotImplemented(err error) bool {
+	return err == errNotImplemented
+}
+
+// ListPaged implements driver.ListPaged.
+func (b *bucket) ListPaged(ctx context.Context, opts *driver.ListOptions) (*driver.ListPage, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	var pageToken string
+	if len(opts.PageToken) > 0 {
+		pageToken = string(opts.PageToken)
+	}
+	pageSize := opts.PageSize
+	if pageSize == 0 {
+		pageSize = defaultPageSize
+	}
+
+	var keys []string
+	for key := range b.blobs {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	// If opts.Delimiter != "", lastPrefix contains the last "directory" key we
+	// added. It is used to avoid adding it again; all files in this "directory"
+	// are collapsed to the single directory entry.
+	var lastPrefix string
+	var result driver.ListPage
+	for _, key := range keys {
+		// Skip keys that don't match the Prefix.
+		if !strings.HasPrefix(key, opts.Prefix) {
+			continue
+		}
+
+		entry := b.blobs[key]
+		obj := &driver.ListObject{
+			Key:     key,
+			ModTime: entry.Attributes.ModTime,
+			Size:    entry.Attributes.Size,
+		}
+
+		// If using Delimiter, collapse "directories".
+		if opts.Delimiter != "" {
+			// Strip the prefix, which may contain Delimiter.
+			keyWithoutPrefix := key[len(opts.Prefix):]
+			// See if the key still contains Delimiter.
+			// If no, it's a file and we just include it.
+			// If yes, it's a file in a "sub-directory" and we want to collapse
+			// all files in that "sub-directory" into a single "directory" result.
+			if idx := strings.Index(keyWithoutPrefix, opts.Delimiter); idx != -1 {
+				prefix := opts.Prefix + keyWithoutPrefix[0:idx+len(opts.Delimiter)]
+				// We've already included this "directory"; don't add it.
+				if prefix == lastPrefix {
+					continue
+				}
+				// Update the object to be a "directory".
+				obj = &driver.ListObject{
+					Key:   prefix,
+					IsDir: true,
+				}
+				lastPrefix = prefix
+			}
+		}
+
+		// If there's a pageToken, skip anything before it.
+		if pageToken != "" && obj.Key <= pageToken {
+			continue
+		}
+
+		// If we've already got a full page of results, set NextPageToken and return.
+		if len(result.Objects) == pageSize {
```

Comment:
That doesn't work though, because you should only set `NextPageToken` when there are more results, and because of the various `continue`s above, it's not until here that you know for sure you've got another result.

Reply:
P.S., this code is largely copied from `fileblob`; added a comment about that.